File: packages/agent-sdk/src/cases.ts
/**
 * Shared case/workspace contracts for the CaseGraph platform.
 *
 * These types define generic case records, case-document references,
 * workflow run tracking, and normalized request/response shapes for the
 * case-centric foundation layer.
 */

import type {
  DocumentId,
  DocumentProcessingStatus,
  IngestionMode,
  IngestionModePreference,
  SourceFileMetadata,
} from "./ingestion";
import type {
  CaseDomainContext,
  CaseTypeTemplateId,
  DomainPackId,
} from "./domains";
import type { CaseTargetPackSelection } from "./target-packs";

export type CaseId = string;
export type CaseTitle = string;
export type CaseCategory = string;
export type WorkflowRunId = string;

export type CaseStatus =
  | "open"
  | "active"
  | "on_hold"
  | "closed"
  | "archived";

export type CaseRunStatus =
  | "created"
  | "running"
  | "completed"
  | "failed"
  | "queued_placeholder"
  | "not_started"
  | "failed_validation"
  | "completed_placeholder";

export type RunInputReferenceType =
  | "case_document"
  | "document"
  | "case"
  | "custom";

export type CaseMetadata = Record<string, unknown>;

export interface TimestampMetadata {
  created_at: string;
  updated_at: string;
}

export interface NormalizedOperationError {
  error_code: string;
  message: string;
  recoverable: boolean;
}

export interface CaseWorkflowBindingMetadata {
  workflow_id: string;
  bound_at: string;
}

export interface CaseRecord {
  case_id: CaseId;
  title: CaseTitle;
  category: CaseCategory | null;
  status: CaseStatus;
  summary: string | null;
  metadata: CaseMetadata;
  domain_context: CaseDomainContext | null;
  workflow_binding: CaseWorkflowBindingMetadata | null;
  target_pack_selection: CaseTargetPackSelection | null;
  timestamps: TimestampMetadata;
}

export interface CaseDocumentReference {
  link_id: string;
  case_id: CaseId;
  document_id: DocumentId;
  source_file: SourceFileMetadata;
  requested_mode: IngestionModePreference | null;
  resolved_mode: IngestionMode | null;
  document_status: DocumentProcessingStatus | null;
  linked_at: string;
}

export interface RunInputReference {
  reference_type: RunInputReferenceType;
  reference_id: string;
  label: string | null;
  metadata: Record<string, unknown>;
}

export interface WorkflowRunOutputPlaceholderMetadata {
  output_available: boolean;
  summary: string | null;
  artifact_refs: string[];
  task_execution_id: string | null;
  task_execution: import("./tasks").TaskExecutionResult | null;
  rag_task_execution: import("./rag").RagTaskExecutionResult | null;
  events: import("./tasks").TaskExecutionEvent[];
}

export interface WorkflowRunRequest {
  workflow_id: string;
  input_references: RunInputReference[];
  linked_document_ids: DocumentId[];
  notes: string | null;
  task_execution?: import("./tasks").TaskExecutionRequest | null;
  rag_task_execution?: import("./rag").RagTaskExecutionRequest | null;
}

export interface WorkflowRunRecord {
  run_id: WorkflowRunId;
  case_id: CaseId;
  workflow_id: string;
  status: CaseRunStatus;
  input_references: RunInputReference[];
  linked_document_ids: DocumentId[];
  output: WorkflowRunOutputPlaceholderMetadata | null;
  events: import("./tasks").TaskExecutionEvent[];
  error: NormalizedOperationError | null;
  notes: string | null;
  timestamps: TimestampMetadata;
}

export interface CreateCaseRequest {
  title: CaseTitle;
  category: CaseCategory | null;
  summary: string | null;
  metadata: CaseMetadata;
  workflow_id: string | null;
  domain_pack_id?: DomainPackId | null;
  case_type_id?: CaseTypeTemplateId | null;
}

export interface UpdateCaseRequest {
  title?: CaseTitle;
  category?: CaseCategory | null;
  status?: CaseStatus;
  summary?: string | null;
  metadata?: CaseMetadata;
  workflow_id?: string | null;
}

export interface LinkCaseDocumentRequest {
  document_id: DocumentId;
}

export interface CaseListResponse {
  cases: CaseRecord[];
}

export interface CaseDocumentListResponse {
  documents: CaseDocumentReference[];
}

export interface WorkflowRunListResponse {
  runs: WorkflowRunRecord[];
}

export interface CaseDetailResponse {
  case: CaseRecord;
  documents: CaseDocumentReference[];
  runs: WorkflowRunRecord[];
}

File: packages/agent-sdk/casegraph_agent_sdk/cases.py
"""Shared case/workspace contracts for the CaseGraph platform."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from casegraph_agent_sdk.domains import CaseDomainContext, CaseTypeTemplateId, DomainPackId
from casegraph_agent_sdk.ingestion import (
    DocumentId,
    DocumentProcessingStatus,
    IngestionMode,
    IngestionModePreference,
    SourceFileMetadata,
)
from casegraph_agent_sdk.rag import RagTaskExecutionRequest, RagTaskExecutionResult
from casegraph_agent_sdk.tasks import TaskExecutionEvent, TaskExecutionRequest, TaskExecutionResult
from casegraph_agent_sdk.target_packs import CaseTargetPackSelection

CaseId = str
CaseTitle = str
CaseCategory = str
WorkflowRunId = str
CaseMetadata = dict[str, Any]

CaseStatus = Literal["open", "active", "on_hold", "closed", "archived"]

CaseRunStatus = Literal[
    "created",
    "running",
    "completed",
    "failed",
    "queued_placeholder",
    "not_started",
    "failed_validation",
    "completed_placeholder",
]

RunInputReferenceType = Literal["case_document", "document", "case", "custom"]


class TimestampMetadata(BaseModel):
    created_at: str
    updated_at: str


class NormalizedOperationError(BaseModel):
    error_code: str
    message: str
    recoverable: bool = False


class CaseWorkflowBindingMetadata(BaseModel):
    workflow_id: str
    bound_at: str


class CaseRecord(BaseModel):
    case_id: CaseId
    title: CaseTitle
    category: CaseCategory | None = None
    status: CaseStatus = "open"
    summary: str | None = None
    metadata: CaseMetadata = Field(default_factory=dict)
    domain_context: CaseDomainContext | None = None
    workflow_binding: CaseWorkflowBindingMetadata | None = None
    target_pack_selection: CaseTargetPackSelection | None = None
    timestamps: TimestampMetadata


class CaseDocumentReference(BaseModel):
    link_id: str
    case_id: CaseId
    document_id: DocumentId
    source_file: SourceFileMetadata
    requested_mode: IngestionModePreference | None = None
    resolved_mode: IngestionMode | None = None
    document_status: DocumentProcessingStatus | None = None
    linked_at: str


class RunInputReference(BaseModel):
    reference_type: RunInputReferenceType
    reference_id: str
    label: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowRunOutputPlaceholderMetadata(BaseModel):
    output_available: bool = False
    summary: str | None = None
    artifact_refs: list[str] = Field(default_factory=list)
    task_execution_id: str | None = None
    task_execution: TaskExecutionResult | None = None
    rag_task_execution: RagTaskExecutionResult | None = None
    events: list[TaskExecutionEvent] = Field(default_factory=list)


class WorkflowRunRequest(BaseModel):
    workflow_id: str
    input_references: list[RunInputReference] = Field(default_factory=list)
    linked_document_ids: list[DocumentId] = Field(default_factory=list)
    notes: str | None = None
    task_execution: TaskExecutionRequest | None = None
    rag_task_execution: RagTaskExecutionRequest | None = None


class WorkflowRunRecord(BaseModel):
    run_id: WorkflowRunId
    case_id: CaseId
    workflow_id: str
    status: CaseRunStatus = "created"
    input_references: list[RunInputReference] = Field(default_factory=list)
    linked_document_ids: list[DocumentId] = Field(default_factory=list)
    output: WorkflowRunOutputPlaceholderMetadata | None = None
    events: list[TaskExecutionEvent] = Field(default_factory=list)
    error: NormalizedOperationError | None = None
    notes: str | None = None
    timestamps: TimestampMetadata


class CreateCaseRequest(BaseModel):
    title: CaseTitle
    category: CaseCategory | None = None
    summary: str | None = None
    metadata: CaseMetadata = Field(default_factory=dict)
    workflow_id: str | None = None
    domain_pack_id: DomainPackId | None = None
    case_type_id: CaseTypeTemplateId | None = None


class UpdateCaseRequest(BaseModel):
    title: CaseTitle | None = None
    category: CaseCategory | None = None
    status: CaseStatus | None = None
    summary: str | None = None
    metadata: CaseMetadata | None = None
    workflow_id: str | None = None


class LinkCaseDocumentRequest(BaseModel):
    document_id: DocumentId


class CaseListResponse(BaseModel):
    cases: list[CaseRecord] = Field(default_factory=list)


class CaseDocumentListResponse(BaseModel):
    documents: list[CaseDocumentReference] = Field(default_factory=list)


class WorkflowRunListResponse(BaseModel):
    runs: list[WorkflowRunRecord] = Field(default_factory=list)


class CaseDetailResponse(BaseModel):
    case: CaseRecord
    documents: list[CaseDocumentReference] = Field(default_factory=list)
    runs: list[WorkflowRunRecord] = Field(default_factory=list)